Run doctests in Julia >= 1.7

fingolfin opened the pull request:
It seems that with the minor adjustments in this PR, doctests pass
in Julia 1.7, 1.8 and 1.9 (but not anymore in 1.6).

If this works, and also with Julia nightly, that may be a good way
to address the problem that currently doctests are run only in Julia
1.6 and thus cannot catch regressions caused by new Julia versions.

The downside is that this means all CI tests will be even slower...

fingolfin commented:
Ah, my local tests only tested the OSCAR doctests. While the `Documentation` CI job also tests AbstractAlgebra doctests, and hence fails... I think that's actually an argument for *not* running doctests in the documentation CI job (we don't want to run AA tests there, this just causes us grief)

benlorenz commented on .github/workflows/CI.yml:
```diff
@@ -61,7 +61,7 @@ jobs:
-          coverage: ${{ matrix.julia-version == '1.6' }}
+          coverage: ${{ matrix.julia-version == '1.8' }}
           depwarn: error
       - name: "Process code coverage"
```
```suggestion
      - name: "Process code coverage"
        if: matrix.julia-version == '1.8'
```
Can you add this to the upload step further down as well?
(This might avoid a bunch of the edits for the codecov comments that often initially claim coverage -75%)

lgoettgens commented:
Another upside of this would be the possibility to debug docs and doctests in a recent julia version, i.e. no need to switch sessions all the time, right? If yes, I am in favor of this!

benlorenz commented on .github/workflows/CI.yml:
```diff
@@ -61,12 +61,14 @@ jobs:
+          coverage: ${{ matrix.julia-version == '1.8' }}
           depwarn: error
       - name: "Process code coverage"
+        if: matrix.julia-version == '1.8
```
```suggestion
        if: matrix.julia-version == '1.8'
```

benlorenz commented on .github/workflows/CI.yml:
```diff
@@ -61,12 +61,14 @@ jobs:
         with:
           directories: src,experimental
       - name: "Upload coverage data to Codecov"
+        if: matrix.julia-version == '1.8
```
```suggestion
        if: matrix.julia-version == '1.8'
```

fingolfin commented:
Hmm, I think the dev docs might mention 1.6 as needed for docs, too. Will look when I am at a computer

fingolfin commented:
I'd be happy to go this way. Question is how much slower the tests really get by running the doctests.

benlorenz commented:
The doctests currently take between 20 and 30 minutes.

I would still prefer to create separate jobs for the doctests.
There was the issue with the coverage upload for the doctests but I think it should also be possible to just upload the coverage results from the doctest job independently and the results will be merged anyway?
This would probably speed up all the linux tests since there are usually some spare runners. For the macos tests it might slow it down slightly since we need to install Oscar each time but the runners are busy most of the time. (We could even do the split only for linux)

Otherwise I am also happy to go forward with this.

fingolfin commented:
> I would still prefer to create separate jobs for the doctests.

Ah, right, I forgot about that -- and I agree with it!

But I won't do it tonight ;-)

fingolfin commented:
(and yeah perhaps we don't split on macOS)

thofma commented:
Will the coverage be merged? If so, I am happy with splitting it off.

lgoettgens commented:
> Will the coverage be merged? If so, I am happy with splitting it off.

According to the [codecov docs](https://docs.codecov.com/docs/merging-reports), yes.

Additionally, there is a parameter that prevents the codecov comment until a certain number of uploads happened (see [here](https://docs.codecov.com/docs/notifications#preventing-notifications-until-after-n-builds). This would be very useful to not get notifications after the doctest action succeded, but only after the normal tests did as well.

thofma commented:
I guess this is only waiting for the separate jobs for the doctests?

fingolfin committed:
Run doctests in Julia >= 1.7

It seems that with the minor adjustments in this PR, doctests pass
in Julia 1.7, 1.8 and 1.9 (but not anymore in 1.6).

fingolfin committed:
Move doctest to seperate CI jobs

... to reduce CI turnaround times. However, don't do this for macOS, as
the extremely limited number of mac runners makes it preferable to keep
the number of mac jobs as low as possible

lgoettgens commented:
Since we now have 3? (ubuntu-1.8-no, ubuntu-1.8-only, macos-1.8-yes) jobs uploading coverage to codecov, I would like to have
```
codecov:
    notify:
        after_n_builds: 3
```
in the `.codecov.yml` to only get the codecov comments, after all three jobs have finished. (see [codecov docu](https://docs.codecov.com/docs/notifications#preventing-notifications-until-after-n-builds))

benlorenz commented on .github/workflows/CI.yml:
```diff
@@ -59,14 +64,29 @@ jobs:
           depwarn: error
+      - name: "Run doctests"
+        if: matrix.doctest != 'no'
+      - run: |
```
```suggestion
        run: |
```

fingolfin committed:
Run the doctests in fully seperate jobs

We loose running doctests on macOS this way, but I think that's an
acceptable compromise: it's unlikely the results for those will differ
on macOS; indeed we mostly test on macOS to verify JLLs work there,
and to catch OS specific regressions on nightly.

If we truly want it, we could re-enable it with some effort, but for
now it doesn't seem worth it to me.

fingolfin commented:
This is in principle ready, but I am not sure whether we really need to / want to run the doctests in four Julia versions. Indeed, perhaps time to remove the 1.7 tests altogether? It makes sense to test...
- on 1.6 as that's LTS;
- on nightly so we get early warning about regression;
- on 1.8 as that's the current stable release
- on 1.9rc as that's the upcoming release.

But there seems little reason to test in 1.7.

Also once 1.9.0 is out, we might want to disable the 1.8 tests.

benlorenz commented on .github/workflows/CI.yml:
```diff
@@ -61,12 +61,65 @@ jobs:
+          - '~1.9.0-0'
+          - 'nightly'
+        os:
+          - ubuntu-latest
```
```suggestion
          - ubuntu-latest
        include:
          - julia-version: '1.8'
            os: macOS-latest
```
I would prefer to keep at least one macos doctest job, I think there is quite a bit of code that is only covered by doctests. Would this require anything other than this `include` section (and another of the very few workers ...)?

fingolfin replied:
I think this patch is sufficient, we'll try. 

But using 50% more macOS workers is painful; I've had multiple occasions in the past weeks were macOS runners were still playing catch up with our various PRs and master updated in the evening after I was already at home :-(.

Originally I tried to do what we discussed above and the macOS runners just run both regular tests and doctests. This worked but I felt the result was simply to hard to understand, debug, and modify. So now I've split the job matrices for the two, and I think the result is quite a bit better, esp. since the job name now makes it clear whether regular tests or doctests are run.

Anyway, in this new setup, it wasn't too bad to make it a bit more ugly again and modify the new macOS doctes runner to also run regular tests... and then drop the "regular test runner" for mac.

Let's see if that works (it rather late, so I wouldn't be surprised if I screwed it up again *sigh*)

benlorenz reviewed (approved):
Apart from my comment regarding the macos doctests this is fine from my point of view, thanks for working on this!
And I agree that we can/should drop 1.7 from both tests and doctests.

fingolfin committed:
Drop CI tests on Julia 1.7

fingolfin committed:
CI: add a macOS doctest runner which also runs regular tests

We do it this way because macOS runners are rather costly

fingolfin committed:
simplify

fingolfin commented:
@benlorenz looking OK to you? Then we could squash-merge it, and afterwards adjust the list of CI tests required by the branch protection rules.

benlorenz reviewed (approved):
Looks good, yes!

There was a crash in the doctests on nightly on the first try but it went through after a restart. The codecov notification count can be added in another PR.
